{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Install Libraries\n",
    "Install information can be found [here](https://satijalab.org/seurat/articles/install.html)\\\n",
    "Vignette [here](https://satijalab.org/seurat/articles/integration_mapping.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "#Seurat parameters\n",
    "reference_data = \"path-to-seurat-object\" # Seurat object for reference data\n",
    "query_data = \"path-to-seurat-object\" # Seurat object for query data\n",
    "\n",
    "genome = \"genome-name\" # either hg38 or mm10\n",
    "\n",
    "normalization_method = \"LogNormalize\"\n",
    "normalization_scale_factor = 10000\n",
    "\n",
    "variable_features_method = \"vst\"\n",
    "variable_features_num = 2000\n",
    "\n",
    "weight.reduction = \"pca\" # Dimensional reduction to use for the weighting anchors.\n",
    "n_dims = 30 # Set of dimensions to use in the anchor weighting procedure. If NULL, the same dimensions that were used to find anchors will be used for weighting.\n",
    "\n",
    "threads = 8\n",
    "prefix = \"prefix\" #project name\n",
    "\n",
    "#Terra specific parameters\n",
    "table_name = \"demux_BH3KTLDMXY\"\n",
    "experiment_name = \"gm12878_fresh_RNA\"\n",
    "\n",
    "#Papermill specific parameters\n",
    "papermill = TRUE\n",
    "\n",
    "#jupyter notebook plot sizes\n",
    "options(repr.plot.width=20, repr.plot.height=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################\n",
    "# For test\n",
    "reference_data = \"../../../ReferenceData/BrainAgingSpatialAtlas_snRNAseq.rds\"\n",
    "query_data = \"../../../QueryData/MouseBrain/SS-PKR-129-192-PLATE1-LEFT-HALF.rna.seurat.filtered_rds.mm10.rds\"\n",
    "genome = \"mm10\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "papermill <- as.logical(papermill)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": [
     "r-libraries"
    ]
   },
   "outputs": [],
   "source": [
    "if (!requireNamespace(\"Seurat\", quietly = TRUE))\n",
    "    install.packages(\"Seurat\")\n",
    "if (!requireNamespace(\"future\", quietly = TRUE))\n",
    "    install.packages(\"future\")\n",
    "if (!requireNamespace(\"logr\", quietly = TRUE))\n",
    "    install.packages(\"logr\")\n",
    "if (!requireNamespace(\"grid\", quietly = TRUE))\n",
    "    install.packages(\"grid\")\n",
    "if (!requireNamespace(\"dplyr\", quietly = TRUE))\n",
    "    install.packages(\"dplyr\")\n",
    "if (!requireNamespace(\"gridExtra\", quietly = TRUE))\n",
    "    install.packages(\"gridExtra\")\n",
    "if (!requireNamespace(\"ggplot2\", quietly = TRUE))\n",
    "    install.packages(\"ggplot2\")\n",
    "if (!require(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "if (!requireNamespace(\"EnsDb.Mmusculus.v79\", quietly = TRUE))\n",
    "    BiocManager::install(\"EnsDb.Mmusculus.v79\")\n",
    "if (!requireNamespace(\"EnsDb.Hsapiens.v86\", quietly = TRUE))\n",
    "    BiocManager::install(\"EnsDb.Hsapiens.v86\")\n",
    "\n",
    "suppressMessages(library(Seurat))\n",
    "suppressMessages(library(future))\n",
    "suppressMessages(library(logr))\n",
    "suppressMessages(library(dplyr))\n",
    "suppressMessages(library(grid))\n",
    "suppressMessages(library(gridExtra))\n",
    "suppressMessages(library(ggplot2))\n",
    "suppressMessages(library(EnsDb.Mmusculus.v79))\n",
    "suppressMessages(library(EnsDb.Hsapiens.v86))\n",
    "\n",
    "future.seed=TRUE\n",
    "plan(\"multisession\", workers = threads)\n",
    "options(\"logr.notes\" = FALSE)\n",
    "options(future.globals.maxSize=10e9)\n",
    "set.seed(1234)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to convert gene ID to symbol\n",
    "create_seurat_obj_with_gene_symbol <- function(obj, genome){\n",
    "\n",
    "    # get gene symbol\n",
    "    if(genome == \"hg38\"){\n",
    "        gene.id <- ensembldb::select(EnsDb.Hsapiens.v86, \n",
    "                                      keys= rownames(obj), \n",
    "                                      keytype = \"GENEID\", \n",
    "                                      columns = c(\"SYMBOL\",\"GENEID\"))\n",
    "    \n",
    "    } else if(genome == \"mm10\"){\n",
    "        gene.id <- ensembldb::select(EnsDb.Mmusculus.v79, \n",
    "                                  keys= rownames(obj), \n",
    "                                  keytype = \"GENEID\", \n",
    "                                  columns = c(\"SYMBOL\",\"GENEID\"))\n",
    "    }\n",
    "    # remove genes with empty symbol\n",
    "    gene.id <- subset(gene.id, gene.id$SYMBOL != \"\")\n",
    "\n",
    "    # make gene symbol unique\n",
    "    gene.id$Unique_SYMBOL <- make.unique(gene.id$SYMBOL, \"\")\n",
    "    \n",
    "    counts <- obj@assays$RNA@counts\n",
    "    counts <- counts[gene.id$GENEID, ]\n",
    "    rownames(counts) <- gene.id$Unique_SYMBOL\n",
    "\n",
    "    obj2 <- CreateSeuratObject(counts = counts, meta.data = obj@meta.data)\n",
    "    \n",
    "    return(obj2)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Function to save plots\n",
    "# plot_filename = paste0(prefix,\".rna.seurat.annotation.plots.\",genome)\n",
    "# dir.create(plot_filename, showWarnings=F)\n",
    "# printPNG <- function(name, plotObject, papermill, wf=22, hf=11){\n",
    "#     filename = paste0(plot_filename,\"/\",prefix,\".rna.seurat.\",name,\".\",genome,\".png\")\n",
    "#     if(papermill){\n",
    "#         ggsave(plot = plotObject, filename = filename, width = wf, height = hf)\n",
    "#     }\n",
    "# }\n",
    "\n",
    "# #Function to create plots\n",
    "# create_plot = function(plot_list, title, subtitle, heights=unit(c(1,1,8), rep(\"in\",3)), width = 7){\n",
    "#     g = c(list(title),list(subtitle),plot_list)\n",
    "#     N = length(plot_list)\n",
    "#     laym = rbind(rep(1,N),rep(2,N),(3:(N+2)))\n",
    "#     widths = unit(rep(width, length(plot_list)), rep(\"in\",length(plot_list)))\n",
    "#     obj = arrangeGrob(grobs=g,layout_matrix=laym, heights=heights, widths = widths)\n",
    "#     return(obj)\n",
    "# }\n",
    "\n",
    "\n",
    "#Create log file\n",
    "logfile <- file.path(paste0(prefix,\".rna.seurat.annotation.logfile.\", genome, \".txt\"))\n",
    "lf <- log_open(logfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"# Reading reference data...\"\n",
      "[1] \"SUCCESSFUL: Reading reference data\"\n"
     ]
    }
   ],
   "source": [
    "# Read reference data\n",
    "tryCatch(\n",
    "    {\n",
    "        log_print(\"# Reading reference data...\")\n",
    "        obj.ref <- readRDS(reference_data)\n",
    "        log_print(\"SUCCESSFUL: Reading reference data\")\n",
    "    \n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Reading reference data\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"# Reading query data...\"\n",
      "[1] \"SUCCESSFUL: Reading query data\"\n"
     ]
    }
   ],
   "source": [
    "# Read query data\n",
    "tryCatch(\n",
    "    {\n",
    "        log_print(\"# Reading query data...\")\n",
    "        obj.query <- readRDS(query_data)\n",
    "        log_print(\"SUCCESSFUL: Reading query data\")\n",
    "    \n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Reading query data\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"# Converting gene id to symbol for reference data\"\n",
      "[1] \"SUCCESSFUL: Converting gene id to symbol for reference data\"\n"
     ]
    }
   ],
   "source": [
    "# Convert gene ID to symbol for reference data\n",
    "tryCatch(\n",
    "    {\n",
    "        log_print(\"# Converting gene id to symbol for reference data\")\n",
    "        obj.ref <- create_seurat_obj_with_gene_symbol(obj = obj.ref, \n",
    "                                                      genome = genome)\n",
    "        log_print(\"SUCCESSFUL: Converting gene id to symbol for reference data\")\n",
    "\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Converting gene id to symbol for reference data\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"# Subseting reference data\"\n",
      "[1] \"SUCCESSFUL: Subseting reference data\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Not all features provided are in this Assay object, removing the following feature(s): Tafa1, Gm32338, BC005561, Nrg1, Gm32647, Gm42418, Gm10754, Tafa2, Gm10419, Gm26871, mt-Co1, 2010300C02Rik, Gmnc, 2700081O15Rik, 4921539H07Rik, Gm33228, Gm20754, Adgrl4, Rflnb, Lhfpl3, March1, Tmem94, Gm10649, mt-Nd1, Gm5127, Epb41l4a, Drd1, Gm28905, Mrm2, Pakap.1, Gm5820, Rmst, C78859, Gm14051, Gm10516, C230072F16Rik, Gm30371, C130073E24Rik, Gm39185, Gm44593, Twnk, Rtl4, Gm867, Gm38413, Gm12367, Gm49969, Gm45356, Gm6260, BC051408, Gm20457, Gm31645, Gm13561, Ints11, Gm14412, Gm11906, Rps6ka2, Dele1, Gm49678, Tmem131l, Gm20125, Gm45459, Gm49164, Sdhaf4, Gm11542, Adgrl2, Gm1604a, Usf3, Ccn2, Gm42196, Gm15587, Gm4890, Phf24, Gm43598, Gm42851, P3h1, Gm30382, Gm27188, Gm26673, Mrvi1, Vxn, Gsdme, Gm33677, Cip2a, Gm20387, 4930547E14Rik, Znrd2, C530008M17Rik, Gm2164, Gm14033, Gm39043, Plpp3, Cramp1l, Gm12689, Gm26621, Gm38560, 4930488L21Rik, Fam71d, 5330417C22Rik, 4933406B17Rik, Gm19325, 9530018F02Rik, Gm47814, Gm37876, Crybg2, 9630013A20Rik, Gm4577, Gm2824, Gm14004, Gm17473, Gm26691, Wdr49, Tmem121b, Gm12371, Cfap157, Gm13403, 4931419H13Rik, Gm10287, Gm16023, Gm42517, Gm20635, Gm20685, Gm44079, AC150683.1, Gm45470, Gm27207, Gm47772, Rpl10-ps3, Colca2, Gm36120, Gm36908, Ccnq, Gm11762, AI463229, Gm31683, Gm8281, Gm34934, Gm534, Gm9748, Gm4675, 6430553K19Rik, Gm36001, Gm34455, AA388235, Gm50061, Gm50321, Plpp6, Rtl8c, Plppr3, Gm16083, Pwwp3a, Gm10801, Lncpint, Armh4, mt-Nd4l, Gm42705, Gm15411, Gm10010, 4933429O19Rik, Gm34237, Gm43154, March11, Gm15631, Gm15706, Gm11149, Gm14168, Ccdc189, Bbof1, Fam49a, B230209E15Rik, Gm12239, mt-Co2, Gm45460, Tedc1, Gm36839, Gm2897, Pakap, Pde2a, Zfhx2os, Adgrf5, Trp53rka, Gm17690, Gm20559, Minar1, Gm43647, Cybc1, Rasgef1b, Ccdc187, Cavin1, Gm45341, Gm29478, Mir9-3hg, Plppr1, Sept1, Spag6l, Gm48508, Rps19, Gm28784, Marc1, Gcn1, Gm14110, Gm14406, Gm40055, Gm13212, Gm17733, Lockd, Arpin, Gm31045, Gm16348, Gm34885, Gm47328, Calhm5, Gm34776, Gm30122, Gm12059, Gm12212, Gm3693, Vmn1r204, Gm48350, BC067074, Gm3488, Gm27153, Gm17597, Gm20621, Gm41396, 4933432I09Rik, Gm49970, Gm32081, Cc2d2b, Cfap299, Gm13322, Gm40123, Gm13054, Cfap73, 6330420H09Rik, Gm27253, Gm36745, Gm48065, Gm17249, Hist1h1e, 4930578N18Rik, Gm14493, Rps14, Ccdc151, Gm35161, Sox2ot, Rex1bd, Gm10602, Sox1ot, Rps25, Gm12353, Gm9899, Gm44618, Gm34567, Gm32061, Gm11491, 4833427G06Rik, 9330199G10Rik, 2700054A10Rik, 2610301B20Rik, Tceal9, Gm6277, Gm49067, Rpl30, Gm16168, Slf2, mt-Atp6, Rpl37, Gm16973, Gm29676, Disp3, Zfp729b, Gm20268, Gm10138, Gm37885, Gm37986, Brd3os, Gm20515, Gm42303, Gm5860, Ttc39aos1, Fzd10os, C530044C16Rik, Gm15624, Gm45620, Gm20501, Gm48293, Eef1akmt3, Gm11696, G630093K05Rik, Gm33045, Gm20712, A230051N06Rik, C030004G16Rik, D330050I16Rik, Gm32341, 6330411D24Rik, Large1, Zup1, Gm10635, Gm12536, 4930474H20Rik, 3110039M20Rik, Gm6145, Arhgap45, Gm13052, Gm5089, Gm5464, Gm15328, Gm29674, Gm10800, Ccdc180, Gm5149, Gm35438.1, Gm31243, Gm13986, Gm19744, Cbarp, Gm39465, Olfr1564, Plppr5, Gm5577, Rpl18a, Rps8, 4930420G21Rik\"\n"
     ]
    }
   ],
   "source": [
    "# Subset reference data\n",
    "tryCatch(\n",
    "    {\n",
    "        log_print(\"# Subseting reference data\")\n",
    "        gene.common <- intersect(rownames(obj.ref), rownames(obj.query))\n",
    "        \n",
    "        log_print(\"SUCCESSFUL: Subseting reference data\")\n",
    "        obj.ref <-  subset(obj.ref, features = gene.common)\n",
    "        obj.query <-  subset(obj.query, features = gene.common)\n",
    "    },\n",
    "    warning = function(conda){\n",
    "        log_print(\"WARNING: Subseting reference data\")\n",
    "        log_print(cond)\n",
    "    }\n",
    "     error = function(cond) {\n",
    "        log_print(\"ERROR: Subseting reference data\")\n",
    "        log_print(cond)\n",
    "    }\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Predict labels for query dataset\n",
    "tryCatch(\n",
    "    {\n",
    "        log_print(\"# Predicting labels for query data\")\n",
    "        \n",
    "        obj.ref <- obj.ref %>%\n",
    "            NormalizeData(verbose = FALSE) %>%\n",
    "            FindVariableFeatures(selection.method = variable_features_method, \n",
    "                                 nfeatures = variable_features_num)\n",
    "        \n",
    "        obj.query <- obj.query %>%\n",
    "            NormalizeData(verbose = FALSE) %>%\n",
    "            FindVariableFeatures(selection.method = variable_features_method,\n",
    "                                 nfeatures = variable_features_num)\n",
    "        \n",
    "        transfer.anchors <- FindTransferAnchors(\n",
    "            reference = obj.ref,\n",
    "            query = obj.query,\n",
    "            reduction = \"cca\",\n",
    "            verbose = FALSE\n",
    "        )\n",
    "        \n",
    "        predictions <- TransferData(anchorset = transfer.anchors, \n",
    "                                    refdata = obj.ref$cell_type,\n",
    "                                    weight.reduction = obj.query[[weight.reduction]],\n",
    "                                    dims = 1:n_dims,\n",
    "                                   verbose = FALSE)\n",
    "        \n",
    "        log_print(\"SUCCESSFUL: Predicting labels for query data\")\n",
    "    },\n",
    "     error = function(cond) {\n",
    "        log_print(\"ERROR: Predicting labels for query data\")\n",
    "        log_print(cond)\n",
    "    }\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_file <- function(path){\n",
    "#     dest <- getwd()\n",
    "#     gsutil_cp(path, dest)\n",
    "#     name <- basename(path)\n",
    "#     return(name)\n",
    "# }\n",
    "\n",
    "# if (!papermill){\n",
    "#     table <- avtable(table_name)\n",
    "#     rna_matrix <- get_file(table$h5_matrix[table[, sprintf('%s_id', table_name)] == experiment_name])\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "violin-plot"
    ]
   },
   "outputs": [],
   "source": [
    "# #Pre-filtered violin QC plots \n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"# Pre-filtered Violin Plot\")\n",
    "    \n",
    "#         #Code start to create pre-filtered rna QC violin plot\n",
    "    \n",
    "#         obj <- VlnPlot(rna, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), ncol = 3, combine = FALSE, pt.size = 0)\n",
    "                \n",
    "#         tg <- textGrob('Pre-filtered RNA QC plots', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "\n",
    "#         plot_list = list(obj[[1]] + scale_y_log10() + NoLegend() + theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=8.5)), \n",
    "#                          obj[[2]] + scale_y_log10() + NoLegend() + theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=8.5)), \n",
    "#                          obj[[3]] + theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=8.5)))\n",
    "        \n",
    "#         obj = create_plot(plot_list, tg, sg)\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"prefiltered_violin\", plotObject = obj, papermill = papermill)\n",
    "    \n",
    "#         #Code end to create violin plot\n",
    "    \n",
    "#         log_print(\"SUCCESSFUL: Pre-filtered Violin Plot\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Pre-filtered Violin Plot\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "feature-plots"
    ]
   },
   "outputs": [],
   "source": [
    "# # Create feature plots\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"# Pre-filtered QC Scatterplots\")\n",
    "    \n",
    "#         #Code start to create feature plots\n",
    "    \n",
    "#         plot1 <- FeatureScatter(rna, feature1 = \"nCount_RNA\", feature2 = \"percent.mt\") + \n",
    "#             theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=10))\n",
    "#         plot2 <- FeatureScatter(rna, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\") +\n",
    "#             theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=10))\n",
    "        \n",
    "#         tg <- textGrob('Pre-filtered QC scatterplots', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "\n",
    "#         plot_list = list(plot1,plot2)\n",
    "#         obj = create_plot(plot_list, tg, sg ,heights = unit(c(1,1,8), rep(\"in\",3)), width = 10)\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"prefiltered_qc_scatterplots\", plotObject = obj, papermill = papermill)\n",
    "    \n",
    "#         #Code end to create feature plots\n",
    "    \n",
    "#         log_print(\"SUCCESSFUL: Pre-filtered QC Scatterplots\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Pre-filtered QC Scatterplots\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Subset Seurat Object by QC filters\n",
    "\n",
    "# rna = tryCatch({\n",
    "#         log_print(\"# Subset Seurat Object\")\n",
    "    \n",
    "#         #Code start to create seurat object\n",
    "    \n",
    "#         #rna <- CreateSeuratObject(counts = data, project = prefix, min.cells = min_cells, min.features = min_features)\n",
    "#         #rna[[\"percent.mt\"]] <- PercentageFeatureSet(rna, pattern = \"^MT-\")\n",
    "        \n",
    "#         rna <- subset(rna, subset = PassQC == TRUE)\n",
    "    \n",
    "#         #write postfiltered .h5 matrix  \n",
    "#         write_dgCMatrix_h5(GetAssayData(object = rna, slot = \"counts\"), \n",
    "#                            cols_are = \"barcodes\", \n",
    "#                            h5_target = paste0(prefix,\".rna.seurat.filtered_matrix.\",genome,\".h5\"), \n",
    "#                            ref_name = prefix)\n",
    "    \n",
    "#         #Code end to create seurat object\n",
    "    \n",
    "#         log_print(\"SUCCESSFUL: Subset Seurat Object\")\n",
    "#         return(rna)\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Subset Seurat Object\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Post-filtered violin QC plots \n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"# Post-filtered Violin Plot\")\n",
    "    \n",
    "#         #Code start to create post-filtered rna QC violin plot\n",
    "        \n",
    "#         obj <- VlnPlot(rna, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), ncol = 3, combine = F, pt.size = 0)\n",
    "#         tg <- textGrob('Post-filtered RNA QC plots', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "\n",
    "#         plot_list = list(obj[[1]] + scale_y_log10() + NoLegend() + theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=10)), \n",
    "#                          obj[[2]] + scale_y_log10() + NoLegend() + theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=10)), \n",
    "#                          obj[[3]] + theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=10)))\n",
    "        \n",
    "#         obj = create_plot(plot_list, tg, sg)\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"postfiltered_violin\", plotObject = obj, papermill = papermill)\n",
    "        \n",
    "#         #Code end to create violin plot\n",
    "    \n",
    "#         log_print(\"SUCCESSFUL: Post-filtered Violin Plot\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Post-filtered Violin Plot\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Create feature plots\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"# Post-filtered QC Scatterplots\")\n",
    "    \n",
    "#         #Code start to create feature plots\n",
    "    \n",
    "#         plot1 <- FeatureScatter(rna, feature1 = \"nCount_RNA\", feature2 = \"percent.mt\") +\n",
    "#             theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=10))\n",
    "#         plot2 <- FeatureScatter(rna, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\") +\n",
    "#             theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=10))\n",
    "        \n",
    "#         tg <- textGrob('Post-filtered QC scatterplots', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "\n",
    "#         plot_list = list(plot1,plot2)\n",
    "#         obj = create_plot(plot_list, tg, sg, heights = unit(c(1,1,8), rep(\"in\",3)), width = 10)\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"postfiltered_qc_scatterplots\", plotObject = obj, papermill = papermill)\n",
    "        \n",
    "#         #Code end to create feature plots\n",
    "    \n",
    "#         log_print(\"SUCCESSFUL: Post-filtered QC Scatterplots\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Post-filtered QC Scatterplots\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Set Seurat obj to NULL if number of barcodes exceed 300k - this is a temp fix until better alternative\n",
    "\n",
    "# rna = tryCatch({\n",
    "#     log_print(\"# Check number of barcodes\")\n",
    "    \n",
    "#         if(nrow(rna@meta.data) > 250000){\n",
    "#             stop(\"Number of barcodes exceeding 250k; terminating\")\n",
    "#         } else{\n",
    "#             log_print(\"SUCCESSFUL: Check number of barcodes\")\n",
    "#         }\n",
    "#         return(rna)\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Check number of barcodes\")\n",
    "#         log_print(cond)\n",
    "#         rna = NULL\n",
    "#         return(rna)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "normalization"
    ]
   },
   "outputs": [],
   "source": [
    "# # Normalization\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"Normalization\")\n",
    "    \n",
    "#         #Code start to normalize\n",
    "    \n",
    "#         rna <- NormalizeData(rna, normalization.method = normalization_method, scale.factor = normalization_scale_factor)\n",
    "    \n",
    "#         #Code end to normalize\n",
    "    \n",
    "#         log_print(\"SUCCESSFUL: Normalization\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Normalization\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "find-variable-features"
    ]
   },
   "outputs": [],
   "source": [
    "# # Find Variable Features\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"Finding Variable Features\")\n",
    "        \n",
    "#         #Code start to find variable features\n",
    "    \n",
    "#         rna <- FindVariableFeatures(rna, selection.method = variable_features_method, nfeatures = variable_features_num)\n",
    "        \n",
    "#         # Identify the 10 most highly variable genes\n",
    "#         top10 <- head(VariableFeatures(rna), 10)\n",
    "\n",
    "#         # plot variable features with and without labels\n",
    "#         plot1 <- VariableFeaturePlot(rna)\n",
    "#         plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)\n",
    "    \n",
    "#         tg <- textGrob('Variable Genes: Mean Expression-by-Variance Scatterplot', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "\n",
    "#         plot_list = list(plot2)\n",
    "#         obj = create_plot(plot_list, tg, sg, width = 12, heights = unit(c(1,1,8), rep(\"in\",3)))\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"variable_genes\", plotObject = obj, papermill = papermill, wf = 14, hf = 12)\n",
    "        \n",
    "#         log_print(\"SUCCESSFUL: Finding Variable Features\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Finding Variable Features\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "scaling"
    ]
   },
   "outputs": [],
   "source": [
    "# # Scaling\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"Scaling\")\n",
    "    \n",
    "#         #Code start to scale\n",
    "    \n",
    "#         all.genes <- rownames(rna)\n",
    "#         rna <- ScaleData(rna, features = all.genes)\n",
    "    \n",
    "#         #Code end to scale\n",
    "    \n",
    "#         log_print(\"SUCCESSFUL: Scaling\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Scaling\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "pca"
    ]
   },
   "outputs": [],
   "source": [
    "# #Principal Component Analysis\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"Principal Component Analysis\")\n",
    "\n",
    "#         # Code start to run PCA    \n",
    "    \n",
    "#         rna <- RunPCA(rna, features = VariableFeatures(object = rna))\n",
    "#         obj <- VizDimLoadings(rna, dims = 1:dim_loadings_dim, reduction = \"pca\")\n",
    "        \n",
    "#         tg <- textGrob('PCA: Dimension Loadings', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "#         plot_list = list(obj[[1]], obj[[2]])\n",
    "#         obj = create_plot(plot_list, tg, sg, width = 10, heights = unit(c(1,1,8), rep(\"in\",3)))\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"pca_dim_loadings\", plotObject = obj, papermill = papermill)\n",
    "            \n",
    "#         # Code end to run PCA  \n",
    "\n",
    "#         log_print(\"SUCCESSFUL: Principal Component Analysis\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Principal Component Analysis\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "pca-plot"
    ]
   },
   "outputs": [],
   "source": [
    "# # PCA plot\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"PCA plot\")\n",
    "\n",
    "#         # Code start to create PCA plot \n",
    "    \n",
    "#         obj <- DimPlot(rna, reduction = \"pca\")\n",
    "    \n",
    "#         tg <- textGrob('PCA Plot', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "#         plot_list = list(obj[[1]])\n",
    "#         obj = create_plot(plot_list, tg, sg, width = 12, heights = unit(c(1,1,8), rep(\"in\",3)))\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"pca\", plotObject = obj, papermill = papermill, wf = 14, hf = 12)\n",
    "        \n",
    "#         # Code end to create PCA plot  \n",
    "\n",
    "#         log_print(\"SUCCESSFUL: PCA plot\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: PCA plot\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "heatmap"
    ]
   },
   "outputs": [],
   "source": [
    "# # Heatmap\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"Heatmap\")\n",
    "\n",
    "#         # Code start to create heatmap \n",
    "    \n",
    "#         obj <- DimHeatmap(rna, dims = heatmap_dim, cells = heatmap_cells, fast = FALSE, balanced = heatmap_balanced)\n",
    "#         tg <- textGrob('Heatmap', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "#         plot_list = list(obj[[1]])\n",
    "#         obj = create_plot(plot_list, tg, sg, width = 12, heights = unit(c(1,1,8), rep(\"in\",3)))\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"heatmap\", plotObject = obj, papermill = papermill, wf = 14, hf = 12)\n",
    "        \n",
    "#         # Code end to create heatmap \n",
    "\n",
    "#         log_print(\"SUCCESSFUL: Heatmap\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Heatmap\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "jackstraw-plot"
    ]
   },
   "outputs": [],
   "source": [
    "# # Jackstraw Plot\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"Jackstraw Plot\")\n",
    "\n",
    "#         # Code start to create jackstraw plot \n",
    "    \n",
    "#         rna <- JackStraw(rna, num.replicate = jackstraw_replicates)\n",
    "#         rna <- ScoreJackStraw(rna, dims = 1:jackstraw_score_dim)\n",
    "#         obj <- JackStrawPlot(rna, dims = 1:jackstraw_plot_dim)\n",
    "    \n",
    "#         tg <- textGrob('Jackstraw Plot', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "#         plot_list = list(obj)\n",
    "#         obj = create_plot(plot_list, tg, sg, width = 10,heights = unit(c(1,1,8), rep(\"in\",3)))\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"jackstraw\", plotObject = obj, papermill = papermill)\n",
    "            \n",
    "#         # Code end to create jackstraw plot \n",
    "\n",
    "#         log_print(\"SUCCESSFUL: Jackstraw Plot\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Jackstraw Plot\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "elbow-plot"
    ]
   },
   "outputs": [],
   "source": [
    "# # Elbow Plot\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"Elbow Plot\")\n",
    "\n",
    "#         # Code start to create elbow plot \n",
    "    \n",
    "#         obj <- ElbowPlot(rna)\n",
    "#         tg <- textGrob('Elbow Plot', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "#         plot_list = list(obj)\n",
    "#         obj = create_plot(plot_list, tg, sg, width = 10,heights = unit(c(1,1,8), rep(\"in\",3)))\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"elbow\", plotObject = obj, papermill = papermill)\n",
    "                \n",
    "#         # Code end to create elbow plot \n",
    "\n",
    "#         log_print(\"SUCCESSFUL: Elbow Plot\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Elbow Plot\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "umap"
    ]
   },
   "outputs": [],
   "source": [
    "# # Run UMAP\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"Run UMAP\")\n",
    "\n",
    "#         # Code start to run umap \n",
    "    \n",
    "#         rna <- FindNeighbors(rna, dims = 1:umap_dim)\n",
    "#         rna <- FindClusters(rna, resolution = umap_resolution)\n",
    "#         rna <- RunUMAP(rna, dims = 1:umap_dim)\n",
    "    \n",
    "#         # Code end to run umap \n",
    "\n",
    "#         log_print(\"SUCCESSFUL: Run UMAP\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Run UMAP\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "umap-plot"
    ]
   },
   "outputs": [],
   "source": [
    "# # UMAP Plot\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"UMAP Plot\")\n",
    "\n",
    "#         # Code start to create UMAP plot \n",
    "    \n",
    "#         obj <- DimPlot(rna, reduction = \"umap\") +              \n",
    "#             theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=10))                                                                 \n",
    "\n",
    "#         tg <- textGrob('UMAP Plot: uwot Clustering', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "#         plot_list = list(obj)\n",
    "#         obj = create_plot(plot_list, tg, sg, width = 12)\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"umap\", plotObject = obj, papermill = papermill, wf = 14, hf = 12)\n",
    "    \n",
    "#         grid.newpage()\n",
    "    \n",
    "#         #obj <- FeaturePlot(object = rna, features = \"nCount_RNA\", reduction = \"umap\")\n",
    "#         df = as.data.frame(rna[[\"umap\"]]@cell.embeddings)\n",
    "#         df$nCount_RNA = rna$nCount_RNA\n",
    "    \n",
    "#         obj <- ggplot(data=df) + geom_point(aes(x = UMAP_1, y = UMAP_2, color = log10(nCount_RNA))) + \n",
    "#         scale_color_viridis_b() + \n",
    "#         theme_bw() + \n",
    "#         theme(plot.title = element_text(hjust = 0.5),panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = \"black\"))+              \n",
    "#         theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=10))                                                                 \n",
    "\n",
    "        \n",
    "#         tg <- textGrob('UMAP Plot: Color by # of RNA molecules', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "#         plot_list = list(obj)\n",
    "#         obj = create_plot(plot_list, tg, sg, width = 12)\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"umap_rna_count\", plotObject = obj, papermill = papermill, wf = 14, hf = 12)\n",
    "    \n",
    "#         grid.newpage()\n",
    "    \n",
    "#         #obj <- FeaturePlot(object = rna, features = \"nFeature_RNA\", reduction = \"umap\") \n",
    "#         df$nFeature_RNA = rna$nFeature_RNA\n",
    "    \n",
    "#         obj <- ggplot(data=df) + geom_point(aes(x = UMAP_1, y = UMAP_2, color = log10(nFeature_RNA))) + \n",
    "#         scale_color_viridis_b() + \n",
    "#         theme_bw() + \n",
    "#         theme(plot.title = element_text(hjust = 0.5),panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = \"black\"))+              \n",
    "#         theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=10))                                                                 \n",
    "\n",
    "        \n",
    "#         tg <- textGrob('UMAP Plot: Color by # of genes detected', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "#         plot_list = list(obj)\n",
    "#         obj = create_plot(plot_list, tg, sg, width = 12)\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"umap_gene_count\", plotObject = obj, papermill = papermill, wf = 14, hf = 12)\n",
    "    \n",
    "#         grid.newpage()\n",
    "    \n",
    "#         obj <- FeaturePlot(object = rna, features = \"percent.mt\", reduction = \"umap\")\n",
    "#         tg <- textGrob('UMAP Plot: Color by mitochondria percent', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "#         sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "#                         nrow(x = rna), \" genes across \", ncol(x = rna), \n",
    "#                         \" barcodes \\n------------------------------------ \\n\"), \n",
    "#                         gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "#         plot_list = list(obj)\n",
    "#         obj = create_plot(plot_list, tg, sg, width = 12)\n",
    "#         grid.draw(obj)\n",
    "#         printPNG(name = \"umap_mito\", plotObject = obj, papermill = papermill, wf = 14, hf = 12)\n",
    "        \n",
    "#         grid.newpage()\n",
    "    \n",
    "#         # Code end to create UMAP plot \n",
    "\n",
    "#         log_print(\"SUCCESSFUL: UMAP Plot\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: UMAP Plot\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "plots-zip"
    ]
   },
   "outputs": [],
   "source": [
    "# #Create final output files\n",
    "\n",
    "# tryCatch({\n",
    "#         log_print(\"# Final output files\")\n",
    "    \n",
    "#         #Code start to create final output files\n",
    "    \n",
    "#         files2zip <- dir(plot_filename, full.names = TRUE)\n",
    "#         zip(zipfile = paste0(plot_filename,\".zip\"), files = files2zip)\n",
    "\n",
    "#         saveRDS(rna, file = paste0(prefix,\".rna.seurat.filtered_rds.\",genome,\".rds\"))\n",
    "\n",
    "#         #Code end to create final output files\n",
    "    \n",
    "#         log_print(\"SUCCESSFUL: Final output files\")\n",
    "#     },\n",
    "#     error = function(cond) {\n",
    "#         log_print(\"ERROR: Final output files\")\n",
    "#         log_print(cond)\n",
    "#     }\n",
    "# )\n",
    "\n",
    "# log_close()"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R [conda env:zl_share_seq]",
   "language": "R",
   "name": "conda-env-zl_share_seq-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
